### PR TITLE
Blackwidow Elite and Mamba Elite support

### PR DESCRIPTION
**Pull request by beranat**
This patch adds full support (custom colors and supported effects) for the following hardware:
* Keyboard Blackwidow Elite
* Mouse Mamba  Elite

### CODE REVIEW

**Comment by lah7** [comment]
Thanks for contributing! This is a bit awkward as your contribution patches the old codebase. An overhaul of new stuff is taking place in the `dev-0.4.0` branch, which got rid of this code in favour of a new implementation but that's _literally_ a _work-in-progress_ ([roadmap](https://polychromatic.app/docs/roadmap)) and (still) isn't ready to commit at the moment.

To avoid conflicts later, I will hold off merging. The pull request will remain open so users with a **BlackWidow Elite** may use your patch (likewise with Blade Stealth '16 in #144) in the interim.

I apologise for this limbo. Once the `dev-0.4.0` branch is on par with `master` in terms of features (last one being editing effects), then `dev-0.4.0` will merge into `master` so the daily builds can restart and be stable enough for new SVG device mapping contributions.

**Comment by magbest** [comment]
i am a bit of a noob. How do i add this patch to my polychromatic software?

**Comment by lah7** [comment]
You could install Polychromatic normally (v0.3.12) and then override the changed files. From the **Terminal**:

    git clone https://github.com/beranat/polychromatic.git
    cd polychromatic
    sudo cp data/mapping/blackwidow-elite-keyboard-layout.svg /usr/share/polychromatic/mapping/
    sudo cp data/pages/profile_editor.js /usr/share/polychromatic/pages/
    sudo cp polychromatic-controller /usr/bin/
    cd ../
    rm -r polychromatic/

**Comment by magbest** [comment]
@lah7 I will try this definetly. Sorry for the late response, I didn't think I was going to get an answer this quick. =)

**Comment by borgdylan** [comment]
I have just got a new blackwidow elite to replace a dmaaged chroma v2. I found out thatmy installed polychromatic does not let me map colours for the space bar, alt gr and media buttons+spinner. How do I apply the previously mentioned patch to my already installed polychromatic?

**Comment by beranat** [comment]
Dear **borgdylan**,

**Comment by borgdylan** [comment]
According to ```lsusb``` it is ```1532:0228 Razer USA, Ltd```. The kernel does not list the fuill product name like it does for my deathadder elite. Polychromatic reports it has firware version 1.0.

**Comment by beranat** [comment]
Bus 004 Device 002: ID 1532:0228 Razer USA, Ltd Razer BlackWidow Elite
Bus 003 Device 002: ID 1532:006c Razer USA, Ltd Razer Mamba Elite

**Comment by beranat** [comment]
Dear **@borgdylan **, I have test at Fedora 31 my keyboard and mouse (the same devices)
Bus 001 Device 018: ID 1532:0228 Razer USA, Ltd Razer BlackWidow Elite
Bus 001 Device 017: ID 1532:006c Razer USA, Ltd Razer Mamba Elite
and have not faced any issues. Have you update openrazer driver too? For Fedora you may get issues with 'plugdev' group (in RedHat's branches it named "input")

**Comment by beranat** [comment]
Merry Christmas and Happy New Year for all

**Comment by samindaperamuna** [comment]
Any update on the pull request?

**Comment by lah7** [comment]
I'm still slowly working on v1.0.0 :turtle:  but as this pull request targets the old codebase (v0.3.12) I don't intend to merge as-is, I'll likely cherry-pick bits and bops when the time is right. Sorry for the wait.

Currently, development v1.0.0 doesn't have the new custom effects implemented yet, so I'm holding off merging anything into `master` since users with the "daily" or `-git` package installed would update into a regressed experience.

As for an ETA... "when it's ready", but hopefully everything for v1.0.0 is finalised no later then the end of Q1 2020.

File: locale/ru/polychromatic-common.po [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+"Content-Transfer-Encoding: 8bit\n"
+"Generated-By: pygettext.py 1.5\n"
+"X-Generator: Poedit 2.2.1\n"
+"Last-Translator: \n"
+"Plural-Forms: nplurals=3; plural=(n%10==1 && n%100!=11 ? 0 : n%10>=2 && n%10<=4 && (n%100<12 || n%100>14) ? 1 : 2);\n"
+"Language: ru\n"
+
+#: pylib/common.py:67
+msgid "Spectrum"
+msgstr "Циклич.переключение"
```
Comment (Arilas) [inline locale/ru/polychromatic-common.po:22]: "Спектр" будет намного лучше. Цветовой спектр есть и в русском языке

File: locale/ru/polychromatic-common.po [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+msgid "Wave"
+msgstr "Волна"
+
+#: pylib/common.py:71
+msgid "Reactive"
+msgstr "Обратная реакция"
+
+#: pylib/common.py:73
+msgid "Breath"
+msgstr "Пульсир.переключение"
```
Comment (Arilas) [inline locale/ru/polychromatic-common.po:34]: Дыхание

File: locale/ru/polychromatic-common.po [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+msgid "Spectrum"
+msgstr "Циклич.переключение"
+
+#: pylib/common.py:69
+msgid "Wave"
+msgstr "Волна"
+
+#: pylib/common.py:71
+msgid "Reactive"
+msgstr "Обратная реакция"
```
Comment (Arilas) [inline locale/ru/polychromatic-common.po:30]: В данном случае, это имеет смысл "Отклик" или "Отзыв", а не обратную реакцию

File: locale/ru/polychromatic-controller.po [hunk 4]
```diff
@@ -0,0 +1,957 @@
...
+msgid "None"
+msgstr "Отключено"
+
+#: polychromatic-controller:481
+msgid "Direction:"
+msgstr "Направление:"
+
+#: polychromatic-controller:482 polychromatic-controller:2143
+msgid "Left"
+msgstr "Влево"
```
Comment (Arilas) [inline locale/ru/polychromatic-controller.po:186]: В данном случае имеется ввиду "Левая сторона" либо "Левый бок", к примеру Левый бок мыши.

File: locale/ru/polychromatic-controller.po [hunk 5]
```diff
@@ -0,0 +1,957 @@
...
+
+#: polychromatic-controller:483 polychromatic-controller:2144
+msgid "Right"
+msgstr "Вправо"
```
Comment (Arilas) [inline locale/ru/polychromatic-controller.po:190]: Так же как и выше, только правый бок

File: locale/ru/polychromatic-controller.po [hunk 6]
```diff
@@ -0,0 +1,957 @@
...
+
+#: polychromatic-controller:484
+msgid "Speed:"
+msgstr "Скорость:"
+
+#: polychromatic-controller:485 polychromatic-controller:2156
+#: polychromatic-controller:2163
+msgid "Random"
+msgstr "Случайный "
```
Comment (Arilas) [inline locale/ru/polychromatic-controller.po:199]: Лишний пробел

File: locale/ru/polychromatic-controller.po [hunk 7]
```diff
@@ -0,0 +1,957 @@
...
+msgid "Disable"
+msgstr "Выключить"
+
+#: polychromatic-controller:497
+msgid "Enable"
+msgstr "Включить"
+
+#: polychromatic-controller:498
+msgid "Disables ALT+F4 and Super key."
+msgstr "Отключить комбинацию клавиш Alt + F4 и кновку Super."
```
Comment (Arilas) [inline locale/ru/polychromatic-controller.po:249]: Опечатка "кнопку"

File: locale/ru/polychromatic-controller.po [hunk 8]
```diff
@@ -0,0 +1,957 @@
...
+#: polychromatic-controller:529
+#, fuzzy
+msgid "Macro Keys Usage"
+msgstr "Использование макро-клавиш"
+
+#: polychromatic-controller:530
+msgid ""
+"You can also use on-the-fly macro recording (provided by the daemon) by "
+"following these steps:"
+msgstr ""
```
Comment (Arilas) [inline locale/ru/polychromatic-controller.po:346]: Оставь оригинальный текст хотя бы в строках дальше

**Comment by Giuseppe499** [comment]
> You could install Polychromatic normally (v0.3.12) and then override the changed files. From the **Terminal**:
> 
> ```
> git clone https://github.com/beranat/polychromatic.git
> cd polychromatic
> sudo cp data/mapping/blackwidow-elite-keyboard-layout.svg /usr/share/polychromatic/mapping/
> sudo cp data/pages/profile_editor.js /usr/share/polychromatic/pages/
> sudo cp polychromatic-controller /usr/bin/
> cd ../
> rm -r polychromatic/
> ```

Hi @lah7 ,

Thank you for your support!

I did what you said and now when i try to launch polychromatic-controller from terminal i get the error:

`Traceback (most recent call last):
  File "/usr/bin/polychromatic-controller", line 38, in <module>
    from platform import linux_distribution
ImportError: cannot import name 'linux_distribution' from 'platform' (/usr/lib/python3.8/platform.py)
`

I think that the problem is that linux_distribution is deprecated and cannot be used in python 3.8. What should I do?

Thank you for your work and contributions!

**Comment by lah7** [comment]
Hi @Giuseppe499,

This branch/fork doesn't have the 0618e107eebda808dbaef98efa8938ea67502692 commit which indeed swaps the deprecated `linux_distribution` with `distro`.

Running these commands should patch this up:

    wget https://github.com/polychromatic/polychromatic/commit/0618e107eebda808dbaef98efa8938ea67502692.patch
    sudo patch -p0 < 0618e107eebda808dbaef98efa8938ea67502692.patch
    # File to patch: /usr/bin/polychromatic-controller

**Comment by Giuseppe499** [comment]
> Hi @Giuseppe499,
> 
> This branch/fork doesn't have the [0618e10](https://github.com/polychromatic/polychromatic/commit/0618e107eebda808dbaef98efa8938ea67502692) commit which indeed swaps the deprecated `linux_distribution` with `distro`.
> 
> Running these commands should patch this up:
> 
> ```
> wget https://github.com/polychromatic/polychromatic/commit/0618e107eebda808dbaef98efa8938ea67502692.patch
> sudo patch -p0 < 0618e107eebda808dbaef98efa8938ea67502692.patch
> # File to patch: /usr/bin/polychromatic-controller
> ```

Hi @lah7 ,

Thank you for your fast reply and for your kind help!

I did what you said and now i can start polychromatic-controller without any error. Unfortunately, I still can't access the features added by this branch for my device (Razer Mamba Elite). Maybe is openrazer 2.8.0 not working correctly with this branch?
What can i do? Should I wait for the 1.0 release or can I still try to make this work?

Thank you again for your help, if useful information is missing please let me know.

**Comment by lah7** [comment]
If the device isn't appearing, then it could be the usual driver installation problems - please see [OpenRazer's troubleshooting guide](https://github.com/openrazer/openrazer/wiki/Troubleshooting).

This branch should work, which is providing graphic maps for these specific devices in the "application profiles" editor. You could wait until v1.0.0 for this feature officially when the _framework_ is in place.

**Comment by Giuseppe499** [comment]
Hi @lah7 ,

The device is appearing so I think drivers works fine. The issue is that when I try to use the "application profiles" editor i can't do anything.
This is what i get when using the "application profiles" editor:
Istantanea_2020-10-26_00-11-44

Furthermore, when i enter in this editor all of the Razer Mamba Elite lights turn off.

**Comment by lah7** [comment]
I'm afraid I can't really advise any further since I'm not the author of this branch, sorry. Running in the Terminal may reveal error details. I know that lights going off is because the "live preview" preference setting is enabled.

**Comment by Giuseppe499** [comment]
@lah7 Thank you for your help and for the useful advice, I will try to further investigate this issue watching if some error comes up in the terminal. If I manage to make it work or if I have some interesting update should I reply on this pull request? (sorry I'm new to GitHub, i don't know exactly what should I do)

**Comment by lah7** [comment]
You're welcome! Please only reply further if it's useful for the author to fix, since this **isn't** an officially supported feature (it's a proposed change, but won't be [merged as-is which is reasoned here](https://github.com/polychromatic/polychromatic/pull/200#issuecomment-482214028)).

Replies send notifications to at least 20+ people watching this repository/issue :bell:

[My focus is on v1.0.0](https://github.com/polychromatic/polychromatic/tree/dev-1.0.0-pyqt5), to consolidate a lot of shortcomings, including custom effects for any kind of supported device. v0.3.12 is finished, since that evolved around the BlackWidow Chroma keyboard. Hopefully v1.0.0 will be out before the end of the year _(finally!)_ and the work here will be incorporated.

**Comment by beranat** [comment]
Request is obsoleted by [Blackwidow Elite and Mamba Elite layouts]((https://github.com/polychromatic/polychromatic/pull/313).

**Pull request closed without merging**